### ISSUE TITLE
Issue with composition (crawling another page) + collection

### ISSUE THREAD

**Issue by dfdeagle47** [lapwinglabs/x-ray#135]
Hello,

I have an error when I'm trying to combine composition and collections.

```javascript
x('http://www.allconferences.com/search/index/showLastConference:1/Category__parent_id:1/sort:start_date/direction:asc',
    '.listing_content',
    [{
      title: '.conferenceHead>h2>a@href',
      items: x('.conferenceHead>h2>a@href', {
        test: '.conferenceTitles'
      })
    }])(function(err, obj) {
    console.log(err, obj);
  });
```

So the idea is that I want to get information for several items (using the selector '.listing_content'). Part of the information of a single item is located on the main page, and another part is located on a specific url (which is found using the selector '.conferenceHead>h2>a@href').

If I don't use a collection (no '[', ']'), it works (but of course I get a single item). However, if I use the collection, the result is always undefined.

The selectors are working correctly because I get the correct results at (index.js; line 199):

```javascript
                  return next(null, compact(out));
```

The variable ```out``` contains the information I want. The issue seems to come from (batch module, index.js; line 145):
```javascript
      if (--pending) next();
```
because ```pending``` is equal to 0 when return the variable ```out```, and ```--pending``` is ```-1```, which is truthy and ```next()``` is called. So this method is called once too many (and the batch.end callback has already been called probably which explains the undefined result).

TL;DR: I have to figure out why next is called too many times.

**Comment by dfdeagle47** [comment]
More minimalistic example:

````javascript
x('https://github.com/lapwinglabs/x-ray',
    '.repository-content',
    [{
      nextPage: x('a#a084b794bc0759e7a6b77810e01874f2-86474699a5d14547da53eeb736e17d35192e3c89@href', {
        authorLink: '.entry-title a.url@href'
      })
    }])(function(err, obj) {
    console.log(err, obj);
  });
```

should yield
```javascript
obj = [{
    nextPage: {
        authorLink: "https://github.com/lapwinglabs"
    }
}]
```

**Comment by dfdeagle47** [comment]
Okay, it was caused by (index.js; line 192):

```javascript
$scope.each(function(i, el) {
```
There was no return statement there so (index.js; line 205):
```javascript
  return next();
```
was called, but the same function ```next``` is called inside the ```$scope.each``` callback, once all the items in the collection have been processed (index.js; line 199):
```javascript
return next(null, compact(out));
```

**Comment by thangngoc89** [comment]
Hi @dfdeagle47 , I'm having this issue too. Could you reopen this to let the maintainers know about this situation here? Anyway, your PR #136 is out of date

**Comment by thangngoc89** [comment]
@matthewmueller I have a simple fix here 
https://github.com/lapwinglabs/x-ray/compare/master...thangngoc89:master

But I'm not too familiar with this repo (only know about this 2 hours ago) so I don't know how to make a proper PR. Any suggestions?

**Comment by apparition47** [comment]
@thangngoc89 , your fix works very well for me and it's very clean. I really want to see this merged in so I opened a PR #200 for you.

**Comment by alexchantastic** [comment]
I've tried @thangngoc89's fix in PR #200 but that doesn't seem to fix the issue I point out here https://github.com/lapwinglabs/x-ray/issues/189#issuecomment-290957902.
